### PR TITLE
Move Xbox Authentication From NMP/NBP To Package

### PR DESCRIPTION
**Pull request by Kashalls**

### CODE REVIEW

File: src/MicrosoftAuthFlow.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+class MicrosoftAuthFlow {
+    constructor(username, cacheDir, options = {}, codeCallback) {
+        this.options = options
+        this.initTokenCaches(username, cacheDir)
+        this.codeCallback = codeCallback
+    }
+
+    initTokenCaches(username, cacheDir = mcDefaultFolderPath) {
```
Comment (extremeheat) [inline src/MicrosoftAuthFlow.js:29]: No need to specify a default folder here, it can be specified in the lib

File: src/MicrosoftAuthFlow.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            const scopes = ['service::user.auth.xboxlive.com::MBI_SSL']
+            this.msa = new LiveTokenManager(this.options.authTitle, scopes, cachePaths.live)
+        } else { // Login with microsoftonline.com
+            const scopes = ['XboxLive.signin', 'offline_access']
+            this.msa = new MsaTokenManager(msalConfig, scopes, cachePaths.msa)
+        }
+
+        const keyPair = crypto.generateKeyPairSync('ec', { namedCurve: 'P-256' })
+        this.xbl = new XboxTokenManager(Authentication.XSTSRelyingParty, keyPair, cachePaths.xbl)
+        this.mca = new MinecraftTokenManager(cachePaths.bed)
```
Comment (extremeheat) [inline src/MicrosoftAuthFlow.js:60]: need a good way to support both editions, not sure if we want to initialize everything here, maybe only based on the passed options

File: src/MicrosoftAuthFlow.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                    return xsts
+                } else {
+                    const xsts = await this.xbl.getXSTSToken(ut)
+                    return xsts
+                }
+            }, () => { this.msa.forceRefresh = true }, 2)
+        }
+    }
+
+    async getMinecraftToken(publicKey) {
```
Comment (extremeheat) [inline src/MicrosoftAuthFlow.js:122]: This can be renamed to getMinecraftBedrockToken() or something similar. Or the game target could be specified in the constructor and token obtained via getFinalToken/getGameToken().

File: src/MicrosoftAuthentication.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * Authenticates with Mincrosoft through user credentials, then
+ * with Xbox Live, Minecraft, checks entitlements and returns profile
+ *
+ * @function
+ * @param {object} client - The client passed to protocol
+ * @param {object} options - Client Options
+ */
+async function authenticatePassword(client, options) {
+    throw Error('Not implemented')
```
Comment (extremeheat) [inline src/MicrosoftAuthentication.js:12]: Only difference for password based flow would be the ms auth part, the xbox auth and part would be the same. So this can be added to msauthflow or a new class which extends it

File: test/authenticatePassword.test.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+/* eslint-env jest */
+
+const { authenticatePassword } = require("../src/MicrosoftAuthentication")
+
+describe('authenticatePassword', () => {
```
Comment (extremeheat) [inline test/authenticatePassword.test.js:5]: we could make a dummy account with a secret to test the auth

File: src/MicrosoftAuthentication.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+* @param {object} client - The client passed to protocol
+* @param {object} options - Client Options
+*/
+async function authenticateDeviceCode(client, options) {
+    try {
+        const flow = new MsAuthFlow(options.username, options.profilesFolder, options, options.onMsaCode)
+
+        const chain = await flow.getMinecraftToken(client.clientX509)
+        // console.log('Chain', chain)
+        await postAuthenticate(client, options, chain)
```
Comment (extremeheat) [inline src/MicrosoftAuthentication.js:29]: The emissions and postAuth stuff should be separate from the lib, where we're just interested in doing the auth.

**Review by extremeheat — CHANGES_REQUESTED** [review]

File: src/MicrosoftAuthFlow.js (hunk 2 above)
Comment (Kashalls) [reply src/MicrosoftAuthFlow.js]: We could probably promisify the function or return a callback once were done authenticating. returns the Xbox profile?

File: src/MicrosoftAuthentication.js (hunk 4 above)
Comment (Kashalls) [reply src/MicrosoftAuthentication.js]: I am thinking we should expose MsAuthFlow and just let the user set a username and password, but if password isnt set just straight go to devicecode.

File: src/MicrosoftAuthFlow.js (hunk 2 above)
Comment (extremeheat) [reply src/MicrosoftAuthFlow.js]: It's already async promise no?

File: src/MicrosoftAuthentication.js (hunk 4 above)
Comment (extremeheat) [reply src/MicrosoftAuthentication.js]: Yeah, that works. The options should be explicit also.

**Comment by Kashalls** [comment]
Alright at this time, the code for implementing the xbox replay is kind of messy. 

I am going to need help testing this. Due to the nature of Microsoft and Password Authentication, Password Authentication is flakey. I have attempted it on 3 Microsoft Accounts and have encountered constant errors on one of them. Otherwise, I think I should just write tests for it real quick and we should be good.

File: README.md [hunk 7]
```diff
@@ -1,21 +1,54 @@
...
+npm install @prismarinejs/xboxlive-auth
+```
 
 ## Usage
 
+Device Code Authentication:
 ```js
-const template = require('prismarine-template')
+const doAuth = async() => {
+    const PrismarineXAuth = require('@prismarinejs/xboxlive-auth');
```
Comment (u9g) [inline README.md:19]: why @prismarinejs ? just xboxlive-auth

Comment (u9g) [reply README.md:19]: + no need to say prismarine just say XAuth

File: index.js [hunk 8]
```diff
@@ -4,6 +4,10 @@ if (typeof process !== 'undefined' && parseInt(process.versions.node.split('.')[
   process.exit(1)
 }
 
-module.exports.helloWorld = function () {
-  console.log('Hello world !')
+module.exports = {
+  Tokens: require('./src/Tokens'),
+  Authflow: require('./src/MicrosoftAuthFlow'),
+
+  Authenticate: require('./src/MicrosoftAuthentication'),
```
Comment (u9g) [inline index.js:11]: is it a typo to have two of these with the change in caps?

File: README.md [hunk 9]
```diff
@@ -1,21 +1,54 @@
...
+    const XSTSToken = await PrismarineXAuth({ username: 'mineflayer@is.cool', password: 'GoCheckItOut!123' });
+    console.log(XSTSToken)
+}
+
+doAuth()
+```
 
-### helloWorld()
+###Expected Response:
```
Comment (u9g) [inline README.md:38]:
```suggestion
### Expected Response:
```

File: package.json [hunk 10]
```diff
@@ -25,7 +25,15 @@
   "homepage": "https://github.com/PrismarineJS/prismarine-template#readme",
   "devDependencies": {
     "jest": "^27.0.4",
-    "prismarine-template": "file:.",
```
Comment (u9g) [inline package.json:28]: maybe add "xboxlive-auth" as a self dep so it would be easier to reference this package in tests? idk if this pr will implement tests though

File: src/MicrosoftAuthFlow.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    } catch (e) {
+      console.log('Failed to open cache dir', e)
+      cache = __dirname
+    }
+
+    const cachePaths = {
+      live: path.join(cache, `./${hash}_live-cache.json`),
+      msa: path.join(cache, `./${hash}_msa-cache.json`),
+      xbl: path.join(cache, `./${hash}_xbl-cache.json`),
+      bed: path.join(cache, `./${hash}_bed-cache.json`)
```
Comment (u9g) [inline src/MicrosoftAuthFlow.js:49]: what is bed? it's not used anywhere else as far as I can tell

File: src/MicrosoftAuthFlow.js [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+          const xsts = await this.xbl.getXSTSToken(ut)
+          return xsts
+        }
+      }, () => { this.msa.forceRefresh = true }, 2)
+    }
+  }
+}
+
+function sha1 (data) {
+  return crypto.createHash('sha1').update(data || '', 'binary').digest('hex')
```
Comment (u9g) [inline src/MicrosoftAuthFlow.js:133]:
```suggestion
  return crypto.createHash('sha1').update(data ?? '', 'binary').digest('hex')
```

File: src/MicrosoftAuthentication.js [hunk 13]
```diff
@@ -0,0 +1,21 @@
+const MicrosoftAuthFlow = require('./MicrosoftAuthFlow')
+
+/**
+* Authenticates to Minecraft with Microsoft.
+*
+* @function
+* @param {object} options - Client Options
```
Comment (u9g) [inline src/MicrosoftAuthentication.js:7]: this isn't how you jsdoc an object, take a look at https://jsdoc.app/tags-typedef.html

File: src/Tokens.js [hunk 14]
```diff
@@ -0,0 +1,544 @@
+const fs = require('fs')
+const path = require('path')
+const crypto = require('crypto')
+
+const msal = require('@azure/msal-node')
+const XboxLiveAuth = require('@xboxreplay/xboxlive-auth')
+const debug = require('debug')('minecraft-protocol')
+const { SmartBuffer } = require('smart-buffer')
+const jose = require('jose-node-cjs-runtime/jwk/from_key_like')
```
Comment (u9g) [inline src/Tokens.js:9]: 1. is it necessary to get the function like this or is it directly exported?
2. instead of using the node runtime, wouldn't it be better to use the general jose package for usage in browser too?

File: src/Tokens.js [hunk 15]
```diff
@@ -0,0 +1,544 @@
...
+    const valid = until > 1000
+    return { valid, until: until, token: token.access_token }
+  }
+
+  getRefreshToken () {
+    const token = this.cache.token
+    if (!token) return
+    const until = new Date(token.obtainedOn + token.expires_in) - Date.now()
+    const valid = until > 1000
+    return { valid, until: until, token: token.refresh_token }
```
Comment (u9g) [inline src/Tokens.js:88]:
```suggestion
    return { valid, until, token: token.refresh_token }
```

File: src/Tokens.js [hunk 16]
```diff
@@ -0,0 +1,544 @@
...
+    this.updateCache(token)
+    return token
+  }
+
+  getAccessToken () {
+    const token = this.cache.token
+    if (!token) return
+    const until = new Date(token.obtainedOn + token.expires_in) - Date.now()
+    const valid = until > 1000
+    return { valid, until: until, token: token.access_token }
```
Comment (u9g) [inline src/Tokens.js:80]:
```suggestion
    return { valid, until, token: token.access_token }
```

File: src/Tokens.js [hunk 17]
```diff
@@ -0,0 +1,544 @@
...
+    if (!rtoken) {
+      throw new Error('Cannot refresh without refresh token')
+    }
+    const refreshTokenRequest = {
+      refreshToken: rtoken.token,
+      scopes: this.scopes
+    }
+
+    return new Promise((resolve, reject) => {
+      this.msalApp.acquireTokenByRefreshToken(refreshTokenRequest).then((response) => {
```
Comment (u9g) [inline src/Tokens.js:258]: this could be simplified with a try-catch-finally

File: src/Tokens.js [hunk 18]
```diff
@@ -0,0 +1,544 @@
...
+    const deviceCodeRequest = {
+      deviceCodeCallback: (resp) => {
+        debug('[msa] device_code response: ', resp)
+        dataCallback(resp)
+      },
+      scopes: this.scopes
+    }
+
+    return new Promise((resolve, reject) => {
+      this.msalApp.acquireTokenByDeviceCode(deviceCodeRequest).then((response) => {
```
Comment (u9g) [inline src/Tokens.js:301]: same as above, try-catch-finally would be much more readable

**Review by u9g — COMMENTED** [review]
overall good, just some nitpicks

File: index.js (hunk 8 above)
Comment (Kashalls) [reply index.js:11]: No, my intention was to solve my OCD and have one capitalized for when I use it, However, I have since fixed that problem. I am going to just keep the lower case one.

File: src/Tokens.js (hunk 14 above)
Comment (Kashalls) [reply src/Tokens.js:9]: From what I understand, thats how we import: https://github.com/panva/jose/blob/HEAD/docs/functions/jwk_from_key_like.fromkeylike.md#function-fromkeylike

Also that is the same package just for CJS not ESM. Thats how the package wants us to use it.

File: README.md [hunk 19]
```diff
@@ -1,21 +1,54 @@
...
+    "XSTSToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiWGJveFJlcGxheS5uZXQifQ.c2UraxPmZ4STYozrjFEW8SBqU0WjnIV0h-jjnfsKtrA",
+    "expiresOn": "2020-04-13T05:43:32.6275675Z"
+}
+```
 
-Prints hello world
+### Parameters
+- options {Object?}
+    - username {string} - Required for authentication.
+    - [password] {string} - Optional, if provided we will attempt password authentication first.
```
[the remark this answers is not in the document]
Comment (Kashalls) [reply README.md:51]: What do you recommend we say then?

[the remark this answers is not in the document]
Comment (extremeheat) [reply README.md:51]: Let it crash and the callee do another request with a different auth flow

File: src/Tokens.js (hunk 14 above)
Comment (extremeheat) [reply src/Tokens.js:9]: The web version of jose is for WebCrypto, we need the node version. Also JwkLike is something that only exists on node.

Comment (Kashalls) [reply src/Tokens.js:9]: > The web version of jose is for WebCrypto, we need the node version. Also KeyLike is something that only exists on node.

We're going to need to figure something out here.
Jest doesn't like the module name: ```Cannot find module 'jose-node-cjs-runtime/jwk/from_key_like' from 'src/Tokens.js```

Comment (extremeheat) [reply src/Tokens.js:9]: You can switch it to mocha

**Comment by extremeheat** [comment]
For the tests, just make sure it gets to the auth step where it asks for a token. Later on if we need we can add a cached token here.

**Comment by rom1504** [comment]
tests are failing

**Comment by Kashalls** [comment]
Thinking we will need to add more tests to test each use case as the Full System needs to be tested but I am not willing to go out and buy another microsoft account to test this

File: src/TokenManagers/JavaTokenManager.js [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+function checkStatus (res) {
+  if (res.ok) { // res.status >= 200 && res.status < 300
+    return res.json()
+  } else {
+    debug('Request fail', res)
+    throw Error(res.statusText)
+  }
+}
+
+class JavaTokenManager {
```
Comment (extremeheat) [inline src/TokenManagers/JavaTokenManager.js:17]: Use MinecraftJava or Mcpc

File: src/MicrosoftAuthFlow.js [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+          return xsts
+        } else {
+          const xsts = await this.xbl.getXSTSToken(ut)
+          return xsts
+        }
+      }, () => { this.msa.forceRefresh = true }, 2)
+    }
+  }
+
+  async getJavaToken () {
```
Comment (extremeheat) [inline src/MicrosoftAuthFlow.js:139]: Same, “java” is ambiguous, avoid it without Minecraft prefix

File: src/MicrosoftAuthFlow.js [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+      this.msa = new LiveTokenManager(this.options.authTitle, scopes, cachePaths.live)
+    } else { // Login with microsoftonline.com
+      const scopes = ['XboxLive.signin', 'offline_access']
+      this.msa = new MsaTokenManager(msalConfig, scopes, cachePaths.msa)
+    }
+
+    const keyPair = crypto.generateKeyPairSync('ec', { namedCurve: 'P-256' })
+    this.xbl = new XboxTokenManager(Authentication.XSTSRelyingParty, keyPair, cachePaths.xbl)
+
+    if (this.options.authTitle) { // Login with bedrock
```
Comment (extremeheat) [inline src/MicrosoftAuthFlow.js:69]: Auth title is used for both Microsoft and Xbox live auth steps. It’s not related to any version of Minecraft. It’s the client ID in the ms auth step, which would probably be a better name. This is always set, this is just an option to customize it since on Bedrock edition, servers get access to extra auth data from the Xbox step which they can choose to verify although vanilla doesn’t. What I was mentioning is keeping a list of known client IDs, but it’s probably not worth it.

File: src/MicrosoftAuthFlow.js [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+    } catch (e) {
+      console.log('Failed to open cache dir', e)
+      cache = __dirname
+    }
+
+    const cachePaths = {
+      live: path.join(cache, `./${hash}_live-cache.json`),
+      msa: path.join(cache, `./${hash}_msa-cache.json`),
+      xbl: path.join(cache, `./${hash}_xbl-cache.json`),
+      mca: path.join(cache, `./${hash}_mca-cache.json`),
```
Comment (extremeheat) [inline src/MicrosoftAuthFlow.js:54]: Both of these use the same file names, they will conflict with each other.

File: README.md [hunk 24]
```diff
@@ -1,21 +1,54 @@
...
+npm install xboxlive-auth
+```
 
 ## Usage
 
+Device Code Authentication:
 ```js
-const template = require('prismarine-template')
+const doAuth = async() => {
+    const XAuth = require('xboxlive-auth');
```
Comment (extremeheat) [inline README.md:19]: Exports and readme do not match

File: src/TokenManagers/BedrockTokenManager.js [hunk 25]
```diff
@@ -0,0 +1,80 @@
+const fs = require('fs')
+const path = require('path')
+const debug = require('debug')('xboxlive-auth')
+const fetch = require('node-fetch')
+
+const { Authentication } = require('../common/Constants')
+
+function checkStatus (res) {
```
Comment (extremeheat) [inline src/TokenManagers/BedrockTokenManager.js:8]: Put this function in a shared file for reuse

File: src/MicrosoftAuthFlow.js [hunk 26]
```diff
@@ -0,0 +1,172 @@
+const fs = require('fs')
+const path = require('path')
+const crypto = require('crypto')
+
+const minecraftFolderPath = require('minecraft-folder-path')
```
Comment (extremeheat) [inline src/MicrosoftAuthFlow.js:5]: Remove this default path from the lib and have it specified by the caller

**Review by extremeheat — CHANGES_REQUESTED** [review]

File: src/MicrosoftAuthFlow.js (hunk 22 above)
Comment (Kashalls) [reply src/MicrosoftAuthFlow.js:72]: How should we identify which one we want?

File: src/MicrosoftAuthFlow.js [hunk 27]
```diff
@@ -0,0 +1,182 @@
...
+    this.username = username
+    this.options = options
+    this.initTokenCaches(username, cacheDir)
+    this.codeCallback = codeCallback
+  }
+
+  initTokenCaches (username, cacheDir) {
+    const hash = sha1(username).substr(0, 6)
+
+    let cache = path.join(cacheDir, 'npm-cache')
```
Comment (extremeheat) [inline src/MicrosoftAuthFlow.js:41]: Remove npm cache related things, just use the provided cache dir

File: src/TokenManagers/MinecraftBedrockTokenManager.js [hunk 28]
```diff
@@ -0,0 +1,72 @@
+const fs = require('fs')
+const path = require('path')
+const debug = require('debug')('xboxlive-auth')
+const fetch = require('node-fetch')
+
+const { Authentication } = require('../common/Constants')
+const { checkStatus } = require('../common/Util')
+
+class BedrockTokenManager {
+  constructor (clientPublicKey, cacheLocation) {
```
Comment (extremeheat) [inline src/TokenManagers/MinecraftBedrockTokenManager.js:10]: Remove first parameter here from the constructor, it’s not used

**Review by extremeheat — COMMENTED** [review]
Overall OK so far

File: README.md [hunk 29]
```diff
@@ -1,21 +1,54 @@
...
+    const XSTSToken = await XAuth.authenticate({ username: 'mineflayer@is.cool', cacheDirectory: './' });
+    console.log(XSTSToken)
+}
 
-template.helloWorld()
+doAuth()
 ```
 
-## API
+Password Authentication (falls back to DeviceCode):
```
[the remark this answers is not in the document]
Comment (Kashalls) [reply README.md:33]: The readme was updated to show examples for fetching pc/pe tokens.

File: src/TokenManagers/MinecraftBedrockTokenManager.js (hunk 28 above)
Comment (Kashalls) [reply src/TokenManagers/MinecraftBedrockTokenManager.js:10]: Are you sure? It looks used to me

File: README.md [hunk 30]
```diff
@@ -1,21 +1,54 @@
...
 [![NPM version](https://img.shields.io/npm/v/prismarine-template.svg)](http://npmjs.com/package/prismarine-template)
 [![Build Status](https://github.com/PrismarineJS/prismarine-template/workflows/CI/badge.svg)](https://github.com/PrismarineJS/prismarine-template/actions?query=workflow%3A%22CI%22)
 [![Discord](https://img.shields.io/badge/chat-on%20discord-brightgreen.svg)](https://discord.gg/GsEFRM8)
 [![Try it on gitpod](https://img.shields.io/badge/try-on%20gitpod-brightgreen.svg)](https://gitpod.io/#https://github.com/PrismarineJS/prismarine-template)
 
-A template repository to make it easy to create new prismarine repo
+Quickly and easily obtain an xbox token to authenticate with Minecraft/Mojang
+
+## Installation
+```shell
```
[the remark this answers is not in the document]
Comment (Kashalls) [reply README.md:10]: What do you want to name it?

[the remark this answers is not in the document]
Comment (u9g) [reply README.md:10]: > I just realized this conflicts with the Xbox replay lib with the same name, let’s try and avoid that

it's not a conflict since they are name scoped

[the remark this answers is not in the document]
Comment (extremeheat) [reply README.md:10]: Yes I get that, but it seems easy to mistake one lib for the other with the same name. Maybe something like xbl-auth or prismarine-xboxlive-auth ? Maybe @rom1504 has some input?

**Review by extremeheat — APPROVED** [review]
Looks good for now

**Comment by Kashalls** [comment]
Maybe because the package will be internal we can nest it under
@prismarinejs?

**Comment by extremeheat** [comment]
Can you add that scope to the package name ?

**Comment by u9g** [comment]
can I ask whats the point of adding contributors if this is going to be an opensource project that I'm sure many will contribute to?

**Comment by Kashalls** [comment]
> can I ask whats the point of adding contributors if this is going to be an opensource project that I'm sure many will contribute to?

It was an internal workflow I had setup with macros and stuff. Trying to simplify my workflow for my internal projects. I reverted the change.

**Review by u9g — CHANGES_REQUESTED** [review]

File: README.md [hunk 31]
```diff
@@ -1,21 +1,68 @@
...
+npm install xboxlive-auth
+```
 
 ## Usage
 
+Device Code Authentication:
 ```js
-const template = require('prismarine-template')
+/* const crypto = require('crypto')
```
[the remark this answers is not in the document]
Comment (Kashalls) [reply README.md:18]: Comments are there to show what you would need to do pass for a minecraft bedrock token.

[the remark this answers is not in the document]
Comment (Kashalls) [reply README.md:18]: I'll rewrite it again.... hold on...

File: README.md [hunk 32]
```diff
@@ -1,21 +1,97 @@
...
 
-Prints hello world
+### Parameters
+Authflow class
+- username {String} - Username for authentication
+- cacheDirectory {String} - Where we will store your tokens
+- options {Object?}
+    - [password] {string} - If passed we will do password based authentication.
+    - cacheDirectory {string} - Where we will store your tokens until you attempt to login next time.
+    - authTitle {string} - Used in authenticating with Switch or Bedrock accounts. See https://github.com/PrismarineJS/xboxlive-auth/src/Constants.js
```
Comment (extremeheat) [inline README.md:96]: No, this is basically required for everything that uses XBL auth. Not setting this will break accounts < 18 years old.

File: README.md [hunk 33]
```diff
@@ -1,21 +1,96 @@
...
+```
 
-Prints hello world
+### Parameters
+Authflow class
+- username {String} - Username for authentication
+- cacheDirectory {String} - Where we will store your tokens
+- options {Object?}
+    - [password] {string} - If passed we will do password based authentication.
+    - [authTitle] {string} - Used in authenticating with Switch or Bedrock accounts. See https://github.com/PrismarineJS/xboxlive-auth/src/Constants.js
```
Comment (extremeheat) [inline README.md:95]: this is not related to bedrock, this should always be set if possible. It is required to switch to live.com auth and do title authentication which is needed for accounts with a date of birth <18 years old. You can update all of the examples to set this.

File: src/MicrosoftAuthFlow.js [hunk 34]
```diff
@@ -0,0 +1,180 @@
...
+      console.log('Failed to open cache dir', e)
+      cache = __dirname
+    }
+
+    const cachePaths = {
+      live: path.join(cache, `./${hash}_live-cache.json`),
+      msa: path.join(cache, `./${hash}_msa-cache.json`),
+      xbl: path.join(cache, `./${hash}_xbl-cache.json`),
+      mca: path.join(cache, `./${hash}_mca-cache.json`),
+      mba: path.join(cache, `./${hash}_mba-cache.json`)
```
Comment (extremeheat) [inline src/MicrosoftAuthFlow.js:56]: please revert the name to `bed` to not break existing cache files

File: package.json [hunk 35]
```diff
@@ -1,17 +1,18 @@
 {
-  "name": "prismarine-template",
-  "version": "1.0.0",
-  "description": "A template repository to make it easy to create new prismarine repo",
+  "name": "@prismarinejs/xboxlive-auth",
+  "version": "0.0.1",
+  "description": "A prismarine utility to authenticate with xbox live and microsoft.",
```
Comment (extremeheat) [inline package.json:4]: remove "prismarine utility"

**Review by extremeheat — CHANGES_REQUESTED** [review]
I'm not sure if the scoped name will go through, but probably should.

Also missing https://github.com/PrismarineJS/bedrock-protocol/blob/master/src/client/titles.js

File: package.json [hunk 36]
```diff
@@ -20,12 +21,22 @@
...
+    "chai-as-promised": "^7.1.1",
+    "mocha": "^9.0.1",
     "standard": "^16.0.1"
+  },
+  "dependencies": {
+    "@azure/msal-node": "^1.1.0",
+    "@xboxreplay/xboxlive-auth": "^3.3.3",
+    "jose-node-cjs-runtime": "^3.13.0",
+    "node-fetch": "^2.6.1",
+    "require-self": "^0.2.3",
```
Comment (extremeheat) [inline package.json:38]: Remove this dep
